src/common/mode/bytes.rs

```diff
@@ -15,250 +15,142 @@
-        );
-
-        Ok(())
-    }
-
-    fn state(&mut self) -> Result<Vec<ScalarValue>> {
-        let values = self.values.take().into_state();
-        let payloads: Vec<ScalarValue> = self
-            .value_counts
-            .take()
-            .get_payloads(&values)
-            .into_iter()
-            .map(|count| match count {
-                Some(c) => ScalarValue::Int64(Some(c)),
-                None => ScalarValue::Int64(None),
-            })
-            .collect();
-
-        let values_list = Arc::new(array_into_list_array_nullable(values));
-        let payloads_list = ScalarValue::new_list_nullable(&payloads, &DataType::Int64);
-
-        Ok(vec![ScalarValue::List(values_list), ScalarValue::List(payloads_list)])
-    }
-
-    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
-        if states.is_empty() {
-            return Ok(());
+            value_counts: HashMap::new(),
+            data_type: data_type.clone(),
         }
-
-        let arr = as_list_array(&states[0])?;
-        let counts = as_primitive_array::<arrow::datatypes::Int64Type>(&states[1])?;
-
-        arr.iter().zip(counts.iter()).try_for_each(|(maybe_list, maybe_count)| {
-            if let (Some(list), Some(count)) = (maybe_list, maybe_count) {
-                // Insert or update the count for each value
-                self.value_counts
-                    .insert_or_update(&list, |_| count, |existing_count| *existing_count += count);
-            }
-            Ok(())
-        })
     }
 
-    fn evaluate(&mut self) -> Result<ScalarValue> {
-        let mut max_index: Option<usize> = None;
-        let mut max_count: i64 = 0;
-
-        let values = self.values.take().into_state();
-        let counts = self.value_counts.take().get_payloads(&values);
-
-        for (i, count) in counts.into_iter().enumerate() {
-            if let Some(c) = count {
-                if c > max_count {
-                    max_count = c;
-                    max_index = Some(i);
-                }
-            }
-        }
-
-        match max_index {
-            Some(index) => {
-                let array = values.as_string::<O>();
-                let mode_value = array.value(index);
-                if mode_value.is_empty() {
-                    Ok(ScalarValue::Utf8(None))
-                } else if O::IS_LARGE {
-                    Ok(ScalarValue::LargeUtf8(Some(mode_value.to_string())))
-                } else {
-                    Ok(ScalarValue::Utf8(Some(mode_value.to_string())))
-                }
-            }
-            None => {
-                if O::IS_LARGE {
-                    Ok(ScalarValue::LargeUtf8(None))
-                } else {
-                    Ok(ScalarValue::Utf8(None))
-                }
+    fn update_counts<'a, V>(&mut self, array: V)
+    where
+        V: ArrayAccessor<Item = &'a str>,
+    {
+        for value in ArrayIter::new(array).flatten() {
+            let key = value.to_string();
```

Comment:
if you can avoid converting the value to a string here it would likely be significantly faster (it would avoid allocating on each row)

So maybe you could only call `to_string()` right before you insert into the map